### ISSUE TITLE
Advanced ROI handling (MR spectroscopy VOI support) - feature request

### ISSUE THREAD

**Issue by ansk** [MRtrix3/mrtrix3#1062]
Dear experts,

I am working with MR spectroscopy datasets (single-voxel MR spectroscopy). The MR spectroscopy voxel is large cuboid VOI with arbitrary size and orientation.

1.
I would need to compute some metrics from the VOI using the 3D anatomical image acquired in the same session, i.e. GM/WM/CSF content of voxel (when providing partial-volume fractions 3D images from `FSL-FAST` or `SPM`) and in case of providing segmentation image (such as `aparc.a2009s+aseg.mgz` from `FreeSurfer`) to provide volumes of particular anatomical compartments in the VOI. 

So my inquiry would be to have a functionality within mrtrix to be able to create cuboid VOI with arbitrary size and orientation and compute summary statistics using this VOI and arbitrary 3D dataset. The option maybe would be to create .mif image defining this VOI (1x1x1 image with large anisotropic voxel size?) to be able to use existing mrstats tools. However, I am not sure how mrstats handles cases when ROI is of differing orientation and voxel size with respect to 3D dataset (i.e. how the voxels lying only partially in the VOI are handled). Ideal would be to account partual-volume voxels properly.

2.
Also, displaying this VOI in `mrview` using ROI tool as intersection with every slice plane would be very nice and handy. 

3.
Another handy tool would be to be able to produce „partial-volume fraction“ image using 3D dataset and this VOI in the same geometry as the provided 3D dataset (this will definitelly belong to mredit framework), where each voxel value would encode whether the image is lying entirely in VOI (value 1), entirely outside VOI (value 0) and for voxels lying only partially in VOI to compute their exact partial volume fraction. This could provide useful partial-volume information for 3D datasets with larger VOIs, such as DWI or fMRI and enable to generate various ROIs for tractography (for example to generate ROI of insula voxels inside spectroscopic VOI using intersection of `aparc.a2009s+aseg.mgz` and this ROI)

The MR spectroscopy dataset is in DICOM provided as class `CSA Non Image`, therefore it can be   peculiar to convert it to `NIfTI` and direct reading of position and orientation from that. 
I tried mrconvert (`mrtrix`), `dcm2nii`, `mri_convert` (`FreeSurfer`) and neither of them was able to read it.

So would be maybe necessary to read them directly from DICOM (for example by using `dcmtk` utilities). I am able to do it myself. So maybe it would suffice to be able to specify this VOI using position of the corner in RAS, 3 direction vectors defining orientation of the edges and length around each vector in mm. 3Rd vector might not be necessary once VOI is always cuboid and can be computed using cross product.

Regards,

Antonin Skoch

**Comment by jdtournier** [comment]
Sorry, I only noticed this post after I'd provided a potential solution on the [corresponding community topic](http://community.mrtrix.org/t/creating-cuboid-roi-with-arbitrary-orientation/1042/3?u=jdtournier). I think that would meet most of your needs - including display within MRView, which works out of the box as far as I can tell. 

I don't think we'll be supporting using completely non-aligned masks within `mrstats` any time soon (it's not a very well-defined problem anyway), so for now the only way of doing this is to regrid your ROI to the same space as the image you want to get your stats from. 

I'm not sure I understand your point 3 - as far as I can tell, this seems like a straightforward multiple operation of your regridded ROI with the partial volume fraction image...? I.e. `mrcalc PVF.mif ROI_regridded.mif -mult PVF_in_ROI.mif` - but I have a feeling I may be getting the wrong end of the stick here... 

As to exporting the information from the spectroscopy DICOMs, I agree this is unlikely to be feasible using standard image converters, but you should be able to glean the relevant information using _MRtrix_'s `dcminfo` command. I don't have spectroscopy data handy, but what I recommend is that you run:
~~~
dcminfo -a data.dcm | less
~~~
then take note of the tags you need. So for an image, that might be:
~~~
[DCM] 0018 0050 DS        2     1008   SliceThickness                         [ 3 ]
[DCM] 0020 0032 DS       38     1552   ImagePositionPatient                   [ -117.5 -126.31355953217 -7.89830493927 ]
[DCM] 0020 0037 DS       12     1598   ImageOrientationPatient                [ 1 0 0 0 1 0 ]
[DCM] 0028 0030 DS       32     1756   PixelSpacing                           [ 2.0982142857143 2.0982142857143 ]
~~~
No idea what the equivalent tags would be for spectroscopy, but hopefully they'll be relatively obvious. Once you've spotted the right tags, note down the corresponding group & element values (the first two numbers after the `[DCM]`), and you can then grab them all using `dcminfo -tag`:
~~~
$ dcminfo data.dcm -tag 0018 0050 -tag 0020 0032 -tag 0020 0037 -tag 0028 0030
3
-117.5 -126.31355953217 -7.89830493927
1 0 0 0 1 0
2.0982142857143 2.0982142857143
~~~
Note the above relies on the [fix I've just committed](https://github.com/MRtrix3/mrtrix3/pull/1052/commits/723e0bc1ac91b076121442d1d5fabe19ed401e51) to the `tag_3.0_RC2` branch a couple of minutes ago... 

Hopefully the above will get you going... Note that you might not find all the direction vectors in the DICOM header, so you may need to compute the z-vector using a cross product of the other two - that's what we do for images anyway. 

Not sure we can really do a lot more than that, this is quite a non-standard case... We do provide the tools you'd need to get this to work, but not in an easily-packaged format unfortunately.

**Comment by ansk** [comment]
Dear @jdtournier,

thank you for the thoroughful assistance. Your comments provided a way how to proceed to handle the problem in mrtrix. In fact, several years before I wrote a `java` binary doing computation of average fraction of GM/WM/CSF in VOI providing segmentation datasets, so I do not have a big problem to get VOI info from DICOM header of MR spectroscopy dataset. But my implementation of binary is not sufficiently flexible and robust and does not offer all functionalities I listed above, that s why I wondered if I can use `mrtrix` tools for that. @Lestropie [post on community forum](http://community.mrtrix.org/t/creating-cuboid-roi-with-arbitrary-orientation/1042/2?u=antonin_skoch) encouraged me to create a github issue for that.

Just to clarify point 3:
I meant to generate "partial-volume-fraction" image of the geometry of 3D input image, with original voxel intensities replaced by following values:
0 if the image voxel is completely outside VOI
1 if the image voxel is completely inside VOI
(0,1) value as a fraction of voxel volume which is inside VOI (i.e. when the voxel is in VOI only in 1/4 of its volume, the voxel value would be 0.25).
So this image is a generalization of VOI, including precise handling of "partial-volume" voxels.

I think that the partial volume matters in this case and could be significant, especially for datasets with large voxel sizes (DWI, fMRI).

I am not sure how complicated this could be for implementation.

**Comment by Lestropie** [comment]
If precision and partial volumes are important, here's an alternative:

- Use your knowledge of the position / orientation / size of your VOI to define vertices corresponding to the eight corners of your cuboid.

- Define 12 triangles to create a closed surface representation of that cuboid (2 triangles per face, obeying the right-hand rule).

- Feed this to `mesh2pve`, using the target image (volume fractions, DWI, ...) as the template.

`mrstats` will interpret a mask image that contains partial volume fractions as a binary mask; if you instead want a *weighted sum / average* within a VOI, you'll need to perform the appropriate operations manually using e.g. `mrcalc`.

**Comment by jdtournier** [comment]
OK, I get your point 3. So this was actually already (almost) possible using `mrtransform` intrinsic oversampling feature. Till now, the amount of oversampling was determined automatically based the relative voxel sizes of the source & destination images, but I've just added an option to control this more explicitly. So once that pull requests makes it out, you'd be able to invoke:
~~~ShellSession
$ mrtransform ROI.mih -template anat.mif ROI_regrid.mif -oversample 3 
~~~

And end up with a better anti-aliased ROI. Higher oversampling factors will of course give you more accurate volume fractions. Hope that does the trick...?

**Comment by ansk** [comment]
Dear @jdtournier,

eventually I am getting again to this issue.
Unfortunately, by `dcminfo` I cannot get the orientation vectors which are encoded in the tag 0029,1110. 
Could you please find time for implementation of decoding of this tag?
Relevant info is here: http://nipy.org/nibabel/dicom/siemens_csa.html
I thought that current `-csa` option could me give this, but unfortunately not.

Regards,

Antonin

**Comment by Lestropie** [comment]
Antonin

The current `master` branch only reads (0029, 1010) and (0029, 1020) as Siemens CSA. In the `dev` branch, it also attempts to read (0029, 1110), (0029, 1120), (0029, 1210) and (0029, 1220) as Siemens CSA. Try compiling that branch and see if it will access the data you're looking for using `dcminfo -csa`.

Rob

**Comment by ansk** [comment]
Dear @Lestropie,

thank you for the prompt response! `dcminfo -csa` in the `dev` branch indeed provides me the information I need.

Antonin

**Comment by ansk** [comment]
Dear experts,
thank you for the support with this. As an example of implementation results see the `mrview` screenshot with GM segmentation map overlaid with single-voxel spectroscopic VOIs centered to insular regions (position, size and orientation of VOI is read from DICOM file).

Regards,

Antonin
![screenshot0000](https://user-images.githubusercontent.com/10279434/36109706-80b6a658-1020-11e8-85ae-f8fcf746bb36.png)

**Comment by Lestropie** [comment]
Thanks for sharing!

- Were you content with the production of a binary ROI for each VOI, rather than a partial volume image? I suspect for volumes of that size, it probably wouldn't make a tremendous difference.

- Is generation of those ROIs something scriptable that you think others may benefit from? If so, you could contribute it to *MRtrix3* :smiley:

**Comment by ansk** [comment]
@Lestropie,

- Using `mrtransform` with `-oversample 4` it gets me quite sufficient partial volume VOI image regridded to segmentation map space:
![screenshot0000](https://user-images.githubusercontent.com/10279434/36146630-904773aa-10b5-11e8-8672-8a28d04d7a73.png)
- I could put here my scripts once they are cleaned up a little and finished. Nevertheless, these are nothing very nice (using `bash` since I am not familiar with `python`).
- To compute correction factor for metabolite concentration of MRS VOI from segmentation maps, I need to compute weighted mean from segmentation map using partial volume VOI image as a per-voxel weight. Could you please suggest, how to compute sum of voxel values in the volume? Surprisingly, I did not find such an option in `mrstats`.

**Comment by ansk** [comment]
Concerning sum of voxel values, I think I could achieve this by executing:
`mrstats image.mif -ignorezero -output count`
`mrstats image.mif -ignorezero -output mean`
and computing product of the values. Am I right?

**Comment by jdtournier** [comment]
Glad to see you've managed to get things working. About computing the sum of values, I'm wondering why you need to compute the sum, when you stated you needed to compute the weighted _mean_...? Nonetheless, if you were after the sum, then yes, the product of the mean by the count would do it (I reckon the `-ignorezero` option is redundant in this case...?). Another (uglier) option might be to use mrmath like this:

    mrmath -axis 0 image.mif sum - | mrmath - -axis 1 sum - | mrmath - -axis 2 sum - | mrstats - -output mean

Maybe one day, we'll make it possible to provide multiple axes to `mrmath` in one go...

Also: I assume your `image.mif` dataset is already the product of your segmentation map by the partial volume VOI image? 

And another thought: I'm surprised you need to rummage around the Siemens CSA private fields to get the information about the VOI. Is this information not stored in standard DICOM tags? As things stand, your pipeline is limited to Siemens data (and data that hasn't been savagely 'anonymised', on top of that). Might not be a problem for you, I guess, but if you ever needed to apply this to data from other scanners...?

**Comment by ansk** [comment]
@jdtournier ,

the reason for need of sum was that I did not know how to compute weighted mean without first computing the per-voxel sum of product of partial volume VOI image and segmentation image and then dividing it by per-voxel sum of the partial volume VOI image.

Without `-ignorezero`, the `count` would give me count of all voxels in file. 

Indeed the VOI position and orientation is not encoded in standard DICOM tags. Siemens is not observing DICOM standard here (at least for VB17 and VE11 versions). There exists DICOM MR spectroscopy SOP class in DICOM standard, but the class Siemens is storing the spectroscopy data is private SOP class CSA non-image.

I am attaching my produced scripts. You can integrate them to `mrtrix3` if you are interested. Any comment/corrections are welcome.
This script reads Siemens CSA non-image single-vogel MR spectroscopy DICOM file and produces `.mih` file with geometry of spectroscopy VOI:
[readVoiPos.txt](https://github.com/MRtrix3/mrtrix3/files/1721833/readVoiPos.txt)

This script reads Siemens CSA non-image single-vogel MR spectroscopy DICOM file, segmentation maps (e.g. from SPM) and computes various correction factors for metabolite concentration estimation using internal water reference:
[segment_svs2.txt](https://github.com/MRtrix3/mrtrix3/files/1721834/segment_svs2.txt)

Antonin

**Comment by Lestropie** [comment]
> Using `mrtransform` with `-oversample 4` it gets me quite sufficient partial volume VOI image regridded to segmentation map space

Yes, that looks quite adequate; I think the issue was that in the previous post, you had loaded those VOI images using the ROI tool rather than the overlay tool, and therefore they appeared binary in the screenshots.

> Could you please suggest, how to compute sum of voxel values in the volume? Surprisingly, I did not find such an option in `mrstats`.

What you're really looking for here is two *volume integrals*: One of the VOI itself, and one of the product of the GM image with the VOI (I find this terminology better than "per-voxel sum", which could be interpreted erroneously). Dividing the latter by the former will give you a mean GM fraction within the VOI, which appropriately takes partial voluming of the VOI into account (which using the "`count`" and "`mean`" outputs from `mrstats` will *not* do).

I've considered on a couple of occasions adding "integral" to the list of outputs provided by `mrstats`, but never convinced myself that it's the right place for it. The `mrmath sum -axis` piping trick is a totally valid alternative, so I don't think it's worth changing `mrstats`.

**Comment by ansk** [comment]
>      What you're really looking for here is two volume integrals:
Yes, _volume integral_ is better term.

> Nonetheless, if you were after the sum, then yes, the product of the mean by the count would do it 

I have changed the code to piping variant of `mrmath`, since product of the mean and count is probably arithmetically less precise (although according to the tests the difference is minimal).

**Comment by jdtournier** [comment]
> Without `-ignorezero`, the count would give me count of all voxels in file.

Sure. What I meant was that if you're after the sum, then as long as the mean was computed from the same number of voxels as the count, then it shouldn't make any difference whether you used all voxels or just those in the ROI - everything outside contributes nothing anyway (it's zero-valued): the mean is 1/_N_ Σ<sub>i</sub> v<sub>i</sub>, and the count is just _N_, so the product will be Σ<sub>i</sub> v<sub>i</sub>, no matter what _N_ was - as long as you're using the same value both times... 

But given subsequent comments, I guess that's a moot point...

> Siemens is not observing DICOM standard here

A major vendor not sticking to the standard - how unexpected... :wink:

> I've considered on a couple of occasions adding "integral" to the list of outputs provided by mrstats, but never convinced myself that it's the right place for it. The mrmath sum -axis piping trick is a totally valid alternative, so I don't think it's worth changing mrstats.

I think in due course, it would make sense to add a `sum` output to `mrstats` - it's easy to do, and as shown here, can come in handy... I also think it would be good to extend `mrmath`'s `-axis` option to allow multiple axes to be specified - I've come across a couple of use cases recently that would have been handled slightly more cleanly with that (not a big deal since it can be handled with just one or two extra stages in the pipe, but still). But these are obviously very low priority... Might be worth filing a couple of issues about this though, just to document them. I'll do that in a sec. 

> I am attaching my produced scripts. You can integrate them to mrtrix3 if you are interested. Any comment/corrections are welcome.

This might be worth thinking about more carefully. I think there's a great deal to be gained from providing an easy framework for people to contribute code snippets based on MRtrix commands, but without them being formally incorporated into _MRtrix3_, with all the maintenance & support obligations this entails. There's a couple of scripts included in the main release at the moment that I'd rather maintain as separate add-ons, for example `gen_scheme` and `convert_bruker`, which are both pretty rough & ready, only added as a convenience to those users that would need it - but within the main release because there isn't really another recognised avenue for sharing them. 

One option would be to use [GitHub gists](https://help.github.com/articles/about-gists/). This would allow users to quickly post scripts for others to use and feedback on, with full version control, etc. The one thing missing would be a way to search for _MRtrix3_-related gists specifically.  But maybe what we can do is use [gist embedding](https://help.github.com/articles/about-gists/#embedding-gists) in a forum post within a new dedicated category, named something like 'user contributions' or something. We could have a description and pinned post requesting that users post a description of what their script is for, how it's to be used, etc. along with the embedded gist. What do you guys reckon?

**Comment by Lestropie** [comment]
That actually looks like the perfect system for this sort of thing. If people use *MRtrix3* in the gist title it'll be discoverable through search, but collating them via the forum would be awesome too. That would give us the ability to provide feedback without the user committing to creating a PR, we could assess on a case-by-case basis whether incorporation into the main repo would be worthwhile, and if not merged it could still be included in a suite of useful tools shared around the *MRtrix3* community.

And as you say, it'd give us somewhere to put those bash scripts we don't really want to support :sweat_smile:

**Comment by jdtournier** [comment]
> If people use MRtrix3 in the gist title it'll be discoverable through search, but collating them via the forum would be awesome too.

Yes, I had a quick look to see whether there was a way to link a gist with a repo, or even just support for hashtags so that a search could find all relevant gists, but no luck. This is why I was suggesting to use a dedicated category on the forum. I guess it would be enough to have 'mrtrix3' in the title though...

**Comment by maxpietsch** [comment]
We could ask users to contribute their code as a link to a gist for instance posted to the forum which we could collect in a submodule or allow more technical users to contribute themselves: http://sunny.today/host-multiple-gists-in-one-git-project/

example: https://github.com/liusally/d3-demo-by-chart-type

**Comment by jdtournier** [comment]
@maxpietsch: the problem with that approach is that it still requires some action on our part to add the gist to the central repo. I'm not too worried about the maintenance overhead of doing this, but I am slightly concerned about the tacit approval that this gives to a 3rd party script that we presumably haven't checked over in any way. If we do manually add it ourselves, this does carry a certain duty to make sure it's kinda OK - even if it's only a perception of such. Or would you just add anything anyone suggests regardless of how buggy it might be? And if you do check things over, where do you set the bar? I just think any interaction from us will open a can of worms... 

This is why I'd rather make the process completely user-driven. If we encourage users to post their scripts as a gist (along with a `README.md`), we can simply ask them to ensure the token `mrtrix3` is in the title. That would allow searching on github, and would be fine for our purposes. But they'll probably also want to advertise this script via the forum anyway. So in many ways, providing a dedicated category explicitly for this serves both as a means for users to alert others to the availability of their scripts, and as a central list of such scripts. So that would be my preferred option.

**Comment by maxpietsch** [comment]
> Or would you just add anything anyone suggests regardless of how buggy it might be?

Yes, that was the idea 🙂  I'd simply name the submodule something along the lines of `unsupported_contributions` to manage expectations. The advantage of a submodule is that it gives some visibility to those snippets which then might trigger more contributions. A github gist search is more effort on the user side, less discoverable and produces results not necessarily intended to be used by the community: https://gist.github.com/search?utf8=%E2%9C%93&q=mrtrix3&ref=simplesearch

**Comment by jdtournier** [comment]
> A github gist search is more effort on the user side, less discoverable...

Well, not if we include a page in the docs with a direct link to the right search. But...

> ... and produces results not necessarily intended to be used by the community

Yes, I was worried this might be the case, thanks for confirming. This was part of the reason I thought a dedicated category on the forum would be better...

> The advantage of a submodule is that it gives some visibility to those snippets which then might trigger more contributions

Yes, I appreciate that, and I agree that visibility is key. The question is how to achieve that. I'm in favour of a forum-based option because it requires zero intervention from us. That's good not just because it's no extra work for us, but also because it's clear we've had no hand in curating these contributions - whether to ensure quality (which would be good) or for any other negative reason (e.g. stifling competitors). It's also immediate - no need for users to get approval from us. I think this kind of actual _and perceived_ independence is quite important to promote these kinds of contributions.

There are other more technical reason why I'm reluctant to go for a submodule-based solution. If this is anything like how the testing data is managed in the main repo (which it should be, it's a submodule), then updates in the upstream repo don't automatically show up in the enclosing repo's submodule. We'd need to fetch, update the submodule, commit and push. Since we're unlikely to check this with any frequency, this will likely require users to notify us every time they make any changes, and then wait for us to actually do something about it. To be fair, I guess we could script this, so it's not necessary a deal-breaker, but hey. The other minor issue is that with this route, users would essentially need to clone the whole thing, even if they're only after one script (although they could just download the one file they were interested in). Finally, assuming this behaves like our testing repo, you'll note that cloning the main repo doesn't actually download the submodule - this only happens when you run the `./run_tests` script, which calls `git submodule update --init`. So getting hold of the script via `git` would actually turn out to be a bit more effort than might have initially been expected - again, all perfectly scriptable, but it all seems like more work than just getting users to post their github gists on the forum...
